internal/sqldb/csv.go

```diff
@@ -0,0 +1,220 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqldb
+
+import (
+	"context"
+	"encoding/csv"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/datacommonsorg/mixer/internal/server/resource"
+	"google.golang.org/api/iterator"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type csvHandle struct {
+	f     io.Reader
+	name  string
+	close func()
+}
+
+type observation struct {
+	entity     string
+	variable   string
+	date       string
+	value      string
+	provenance string
+}
+
+type triple struct {
+	subjectID   string
+	predicate   string
+	objectID    string
+	objectValue string
+}
+
+// Get csv file handle for observation csv and triples csv.
+// Make sure to close the file returned from this function.
+func listCSVFiles(dir string) ([]*csvHandle, []*csvHandle, error) {
+	var obsCSVs []*csvHandle
+	var tripleCSVs []*csvHandle
+	if bucketName, objectPrefix, ok := parseGCSPath(dir); ok {
+		// Read from GCS
+		ctx := context.Background()
+		client, err := storage.NewClient(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		defer client.Close()
+		bucket := client.Bucket(bucketName)
+		query := &storage.Query{
+			Prefix: objectPrefix,
+		}
+		it := bucket.Objects(ctx, query)
+		for {
+			objAttrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return nil, nil, err
+			}
+			// Check if it's a CSV file
+			if strings.HasSuffix(objAttrs.Name, ".csv") {
+				rc, err := bucket.Object(objAttrs.Name).NewReader(ctx)
+				if err != nil {
+					return nil, nil, err
+				}
+				parts := strings.Split(objAttrs.Name, "/")
+				fileName := parts[len(parts)-1]
+				if strings.HasPrefix(fileName, "observations") {
+					obsCSVs = append(
+						obsCSVs,
+						&csvHandle{
+							f:     rc,
+							name:  objAttrs.Name,
+							close: func() { rc.Close() },
+						},
+					)
+				} else if strings.HasPrefix(fileName, "triples") {
+					tripleCSVs = append(
+						tripleCSVs,
+						&csvHandle{
+							f:     rc,
+							name:  objAttrs.Name,
+							close: func() { rc.Close() },
+						},
+					)
+				}
+			}
+		}
+	} else {
+		// Read from local files
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, file := range files {
+			if fName := file.Name(); strings.HasSuffix(fName, ".csv") {
+				f, err := os.Open(filepath.Join(dir, fName))
+				if err != nil {
+					return nil, nil, err
+				}
+				if strings.HasPrefix(fName, "observations") {
```

Comment:
There's some duplication that can be avoided by creating an abstraction for local and GCS files. But ok to go with what we have for now since this code will be deleted once RSI writes to sqlite natively.